File: tools/ghidra/GhidraQuery.py
# GhidraQuery.py - Parameterized headless query script for BioshockHD.exe
# Compatible with both Jython and PyGhidra (Ghidra 11.x / 12.x).
#
# Invoked by re_tool.py via:
#   analyzeHeadless <proj> <name> -process BioshockHD.exe -noanalysis \
#       -scriptPath <dir> -postScript GhidraQuery.py <command> <arg> <outfile>
#
# Commands (each writes JSON to <outfile>):
#   decompile <funcNameOrAddr> <outfile>   - decompiled C of a function
#   func      <namePattern>    <outfile>   - list functions matching substring
#   struct    <structName>     <outfile>   - dump a struct/datatype layout
#   xref      <symbolName>     <outfile>   - callers / references to a symbol
#   search    <asciiString>    <outfile>   - find ASCII string in memory + xrefs
#   strings   <substr>         <outfile>   - list defined strings containing substr
#   data      <addr> <count>   <outfile>   - hex dump <count> bytes at <addr>
#
# All results are JSON so re_tool.py can parse and pretty-print them.

import json

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

program = getCurrentProgram()
listing = program.getListing()
memory = program.getMemory()
af = program.getAddressFactory()
fm = program.getFunctionManager()
st = program.getSymbolTable()
refMgr = program.getReferenceManager()
monitor = ConsoleTaskMonitor()


def write_json(path, obj):
    f = open(path, "w")
    try:
        f.write(json.dumps(obj, indent=2))
    finally:
        f.close()
    print("[GhidraQuery] wrote {}".format(path))


def resolve_addr(s):
    try:
        return af.getAddress(s)
    except:
        return None


def find_function(name_or_addr):
    # Try by exact name first
    funcs = []
    it = fm.getFunctions(True)
    for f in it:
        if f.getName() == name_or_addr:
            return f
    # Try by address
    addr = resolve_addr(name_or_addr)
    if addr is not None:
        f = fm.getFunctionContaining(addr)
        if f is not None:
            return f
    return None


def decompile_func(func):
    decomp = DecompInterface()
    decomp.openProgram(program)
    try:
        res = decomp.decompileFunction(func, 60, monitor)
        if res is not None and res.getDecompiledFunction() is not None:
            return res.getDecompiledFunction().getC()
        return None
    finally:
        decomp.dispose()


def cmd_decompile(arg, outfile):
    func = find_function(arg)
    if func is None:
        write_json(outfile, {"error": "function not found: {}".format(arg)})
        return
    code = decompile_func(func)
    write_json(outfile, {
        "name": func.getName(),
        "entry": str(func.getEntryPoint()),
        "signature": func.getPrototypeString(False, False),
        "code": code if code else "<decompile failed>"
    })


def cmd_func(pattern, outfile):
    results = []
    pl = pattern.lower()
    it = fm.getFunctions(True)
    for f in it:
        if pl in f.getName().lower():
            results.append({
                "name": f.getName(),
                "entry": str(f.getEntryPoint()),
                "signature": f.getPrototypeString(False, False)
            })
            if len(results) >= 400:
                break
    write_json(outfile, {"pattern": pattern, "count": len(results), "functions": results})


def dump_datatype(dt):
    info = {"name": dt.getName(), "length": dt.getLength()}
    try:
        comps = []
        for i in range(dt.getNumComponents()):
            c = dt.getComponent(i)
            comps.append({
                "offset": c.getOffset(),
                "length": c.getLength(),
                "type": c.getDataType().getName(),
                "field": c.getFieldName() if c.getFieldName() else ""
            })
        info["components"] = comps
    except:
        info["components"] = []
    return info


def cmd_struct(name, outfile):
    dtm = program.getDataTypeManager()
    matches = []
    it = dtm.getAllStructures()
    for dt in it:
        if dt.getName() == name or name.lower() in dt.getName().lower():
            matches.append(dump_datatype(dt))
            if len(matches) >= 20:
                break
    write_json(outfile, {"query": name, "count": len(matches), "structs": matches})


def cmd_xref(name, outfile):
    results = []
    syms = st.getGlobalSymbols(name)
    for sym in syms:
        addr = sym.getAddress()
        refs = refMgr.getReferencesTo(addr)
        callers = []
        for r in refs:
            from_addr = r.getFromAddress()
            f = fm.getFunctionContaining(from_addr)
            callers.append({
                "from": str(from_addr),
                "func": f.getName() if f is not None else "<none>",
                "type": str(r.getReferenceType())
            })
        results.append({"symbol": name, "addr": str(addr), "xrefs": callers})
    write_json(outfile, {"query": name, "symbols": results})


def cmd_search(text, outfile):
    results = []
    needle = bytearray(text.encode("ascii"))
    for block in memory.getBlocks():
        if not block.isInitialized():
            continue
        start = block.getStart()
        end = block.getEnd()
        size = end.subtract(start) + 1
        if size > 80000000:
            continue
        buf = bytearray(size)
        try:
            block.getBytes(start, buf)
        except:
            continue
        idx = 0
        s = bytes(buf)
        nd = bytes(needle)
        while True:
            pos = s.find(nd, idx)
            if pos < 0:
                break
            addr = start.add(pos)
            entry = {"addr": str(addr), "block": block.getName()}
            # xrefs to this address
            refs = refMgr.getReferencesTo(addr)
            xr = []
            for r in refs:
                fa = r.getFromAddress()
                f = fm.getFunctionContaining(fa)
                xr.append({"from": str(fa), "func": f.getName() if f is not None else "<none>"})
            entry["xrefs"] = xr
            results.append(entry)
            idx = pos + 1
            if len(results) >= 200:
                break
        if len(results) >= 200:
            break
    write_json(outfile, {"query": text, "count": len(results), "hits": results})


def cmd_strings(substr, outfile):
    results = []
    sl = substr.lower()
    di = listing.getDefinedData(True)
    for d in di:
        try:
            val = d.getValue()
            if val is None:
                continue
            sval = str(val)
            if sl in sval.lower():
                results.append({"addr": str(d.getAddress()), "value": sval})
                if len(results) >= 300:
                    break
        except:
            continue
    write_json(outfile, {"query": substr, "count": len(results), "strings": results})


def cmd_data(addr_str, count_str, outfile):
    addr = resolve_addr(addr_str)
    if addr is None:
        write_json(outfile, {"error": "bad address: {}".format(addr_str)})
        return
    count = int(count_str)
    buf = bytearray(count)
    n = memory.getBytes(addr, buf)
    hexstr = " ".join("{:02x}".format(b & 0xff) for b in buf[:n])
    write_json(outfile, {"addr": str(addr), "count": n, "hex": hexstr})


def main():
    args = getScriptArgs()
    if len(args) < 2:
        print("[GhidraQuery] usage: <command> <arg...> <outfile>")
        return
    cmd = args[0]
    if cmd == "decompile":
        cmd_decompile(args[1], args[2])
    elif cmd == "func":
        cmd_func(args[1], args[2])
    elif cmd == "struct":
        cmd_struct(args[1], args[2])
    elif cmd == "xref":
        cmd_xref(args[1], args[2])
    elif cmd == "search":
        cmd_search(args[1], args[2])
    elif cmd == "strings":
        cmd_strings(args[1], args[2])
    elif cmd == "data":
        cmd_data(args[1], args[2], args[3])
    else:
        print("[GhidraQuery] unknown command: {}".format(cmd))


main()

File: tools/ghidra/GhidraQuery.java
// GhidraQuery.java - Parameterized headless query script for BioshockHD.exe
//
// Java GhidraScript so it runs under standard analyzeHeadless.bat with no
// PyGhidra / Jython dependency. Invoked by re_tool.py:
//
//   analyzeHeadless <proj> <name> -process BioshockHD.exe -noanalysis \
//       -scriptPath <dir> -postScript GhidraQuery.java <command> <arg..> <outfile>
//
// Commands (each writes JSON to <outfile>):
//   decompile <funcNameOrAddr> <outfile>
//   func      <namePattern>    <outfile>
//   struct    <structName>     <outfile>
//   xref      <symbolName>     <outfile>
//   search    <asciiString>    <outfile>
//   strings   <substr>         <outfile>
//   data      <addr> <count>   <outfile>

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.data.DataType;
import ghidra.program.model.data.DataTypeComponent;
import ghidra.program.model.data.Structure;
import ghidra.program.model.data.DataTypeManager;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.mem.Memory;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceManager;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolTable;

import java.io.FileWriter;
import java.util.Iterator;

public class GhidraQuery extends GhidraScript {

    private static String esc(String s) {
        if (s == null) return "";
        StringBuilder b = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  b.append("\\\""); break;
                case '\\': b.append("\\\\"); break;
                case '\n': b.append("\\n");  break;
                case '\r': b.append("\\r");  break;
                case '\t': b.append("\\t");  break;
                default:
                    if (c < 0x20) b.append(String.format("\\u%04x", (int) c));
                    else b.append(c);
            }
        }
        return b.toString();
    }

    private void writeOut(String path, String json) throws Exception {
        FileWriter w = new FileWriter(path);
        try { w.write(json); } finally { w.close(); }
        println("[GhidraQuery] wrote " + path);
    }

    private Function findFunction(String nameOrAddr) {
        FunctionManager fm = currentProgram.getFunctionManager();
        FunctionIterator it = fm.getFunctions(true);
        while (it.hasNext()) {
            Function f = it.next();
            if (f.getName().equals(nameOrAddr)) return f;
        }
        try {
            Address a = currentProgram.getAddressFactory().getAddress(nameOrAddr);
            if (a != null) {
                Function f = fm.getFunctionContaining(a);
                if (f != null) return f;
            }
        } catch (Exception e) { /* ignore */ }
        return null;
    }

    private void cmdDecompile(String arg, String out) throws Exception {
        Function f = findFunction(arg);
        if (f == null) { writeOut(out, "{\"error\":\"function not found: " + esc(arg) + "\"}"); return; }
        DecompInterface di = new DecompInterface();
        di.openProgram(currentProgram);
        String code = "<decompile failed>";
        try {
            DecompileResults r = di.decompileFunction(f, 60, monitor);
            if (r != null && r.getDecompiledFunction() != null)
                code = r.getDecompiledFunction().getC();
        } finally { di.dispose(); }
        StringBuilder b = new StringBuilder();
        b.append("{\"name\":\"").append(esc(f.getName())).append("\",");
        b.append("\"entry\":\"").append(esc(f.getEntryPoint().toString())).append("\",");
        b.append("\"signature\":\"").append(esc(f.getPrototypeString(false, false))).append("\",");
        b.append("\"code\":\"").append(esc(code)).append("\"}");
        writeOut(out, b.toString());
    }

    private void cmdFunc(String pattern, String out) throws Exception {
        String pl = pattern.toLowerCase();
        FunctionManager fm = currentProgram.getFunctionManager();
        FunctionIterator it = fm.getFunctions(true);
        StringBuilder b = new StringBuilder();
        b.append("{\"pattern\":\"").append(esc(pattern)).append("\",\"functions\":[");
        int n = 0;
        while (it.hasNext()) {
            Function f = it.next();
            if (f.getName().toLowerCase().contains(pl)) {
                if (n > 0) b.append(",");
                b.append("{\"name\":\"").append(esc(f.getName())).append("\",");
                b.append("\"entry\":\"").append(esc(f.getEntryPoint().toString())).append("\",");
                b.append("\"signature\":\"").append(esc(f.getPrototypeString(false, false))).append("\"}");
                n++;
                if (n >= 400) break;
            }
        }
        b.append("],\"count\":").append(n).append("}");
        writeOut(out, b.toString());
    }

    private void cmdStruct(String name, String out) throws Exception {
        DataTypeManager dtm = currentProgram.getDataTypeManager();
        Iterator<Structure> it = dtm.getAllStructures();
        String nl = name.toLowerCase();
        StringBuilder b = new StringBuilder();
        b.append("{\"query\":\"").append(esc(name)).append("\",\"structs\":[");
        int n = 0;
        while (it.hasNext()) {
            Structure dt = it.next();
            if (dt.getName().equals(name) || dt.getName().toLowerCase().contains(nl)) {
                if (n > 0) b.append(",");
                b.append("{\"name\":\"").append(esc(dt.getName())).append("\",");
                b.append("\"length\":").append(dt.getLength()).append(",\"components\":[");
                DataTypeComponent[] comps = dt.getComponents();
                for (int i = 0; i < comps.length; i++) {
                    if (i > 0) b.append(",");
                    DataTypeComponent c = comps[i];
                    String fn = c.getFieldName() == null ? "" : c.getFieldName();
                    b.append("{\"offset\":").append(c.getOffset()).append(",");
                    b.append("\"length\":").append(c.getLength()).append(",");
                    b.append("\"type\":\"").append(esc(c.getDataType().getName())).append("\",");
                    b.append("\"field\":\"").append(esc(fn)).append("\"}");
                }
                b.append("]}");
                n++;
                if (n >= 20) break;
            }
        }
        b.append("],\"count\":").append(n).append("}");
        writeOut(out, b.toString());
    }

    private void cmdXref(String name, String out) throws Exception {
        SymbolTable st = currentProgram.getSymbolTable();
        ReferenceManager rm = currentProgram.getReferenceManager();
        FunctionManager fm = currentProgram.getFunctionManager();
        StringBuilder b = new StringBuilder();
        b.append("{\"query\":\"").append(esc(name)).append("\",\"symbols\":[");
        int sn = 0;
        for (Symbol sym : st.getGlobalSymbols(name)) {
            if (sn > 0) b.append(",");
            Address addr = sym.getAddress();
            b.append("{\"symbol\":\"").append(esc(name)).append("\",");
            b.append("\"addr\":\"").append(esc(addr.toString())).append("\",\"xrefs\":[");
            int rn = 0;
            for (Reference r : rm.getReferencesTo(addr)) {
                if (rn > 0) b.append(",");
                Address fa = r.getFromAddress();
                Function f = fm.getFunctionContaining(fa);
                b.append("{\"from\":\"").append(esc(fa.toString())).append("\",");
                b.append("\"func\":\"").append(esc(f == null ? "<none>" : f.getName())).append("\",");
                b.append("\"type\":\"").append(esc(r.getReferenceType().toString())).append("\"}");
                rn++;
                if (rn >= 200) break;
            }
            b.append("]}");
            sn++;
        }
        b.append("]}");
        writeOut(out, b.toString());
    }

    private void cmdSearch(String text, String out) throws Exception {
        Memory mem = currentProgram.getMemory();
        ReferenceManager rm = currentProgram.getReferenceManager();
        FunctionManager fm = currentProgram.getFunctionManager();
        byte[] ascii = text.getBytes("US-ASCII");
        byte[] utf16 = text.getBytes("UTF-16LE");
        byte[][] needles = { ascii, utf16 };
        String[] encs = { "ascii", "utf16le" };
        int maxNeedle = Math.max(ascii.length, utf16.length);
        StringBuilder b = new StringBuilder();
        b.append("{\"query\":\"").append(esc(text)).append("\",\"hits\":[");
        int hits = 0;
        final int CHUNK = 8 * 1024 * 1024;     // 8 MB window
        final int OVERLAP = maxNeedle - 1;     // so matches spanning chunk edges aren't missed
        for (MemoryBlock block : mem.getBlocks()) {
            if (!block.isInitialized()) continue;
            long size = block.getSize();
            long off = 0;
            while (off < size && hits < 200) {
                int readLen = (int) Math.min((long) CHUNK + OVERLAP, size - off);
                byte[] buf = new byte[readLen];
                Address chunkStart = block.getStart().add(off);
                int got;
                try { got = block.getBytes(chunkStart, buf); }
                catch (Exception e) { break; }
                for (int i = 0; i < got; i++) {
                    int encIdx = -1;
                    for (int ni = 0; ni < needles.length; ni++) {
                        byte[] needle = needles[ni];
                        if (i + needle.length > got) continue;
                        boolean match = true;
                        for (int j = 0; j < needle.length; j++) {
                            if (buf[i + j] != needle[j]) { match = false; break; }
                        }
                        if (match) { encIdx = ni; break; }
                    }
                    if (encIdx < 0) continue;
                    Address addr = chunkStart.add(i);
                    if (hits > 0) b.append(",");
                    b.append("{\"addr\":\"").append(esc(addr.toString())).append("\",");
                    b.append("\"enc\":\"").append(encs[encIdx]).append("\",");
                    b.append("\"block\":\"").append(esc(block.getName())).append("\",\"xrefs\":[");
                    int rn = 0;
                    for (Reference r : rm.getReferencesTo(addr)) {
                        if (rn > 0) b.append(",");
                        Address fa = r.getFromAddress();
                        Function f = fm.getFunctionContaining(fa);
                        b.append("{\"from\":\"").append(esc(fa.toString())).append("\",");
                        b.append("\"func\":\"").append(esc(f == null ? "<none>" : f.getName())).append("\"}");
                        rn++;
                    }
                    b.append("]}");
                    hits++;
                    if (hits >= 200) break;
                }
                off += CHUNK; // advance by CHUNK (OVERLAP keeps edge matches)
            }
            if (hits >= 200) break;
        }
        b.append("],\"count\":").append(hits).append("}");
        writeOut(out, b.toString());
    }

    private void cmdStrings(String substr, String out) throws Exception {
        Listing listing = currentProgram.getListing();
        String sl = substr.toLowerCase();
        DataIterator di = listing.getDefinedData(true);
        StringBuilder b = new StringBuilder();
        b.append("{\"query\":\"").append(esc(substr)).append("\",\"strings\":[");
        int n = 0;
        while (di.hasNext()) {
            Data d = di.next();
            Object v = d.getValue();
            if (v == null) continue;
            String sv = v.toString();
            if (sv.toLowerCase().contains(sl)) {
                if (n > 0) b.append(",");
                b.append("{\"addr\":\"").append(esc(d.getAddress().toString())).append("\",");
                b.append("\"value\":\"").append(esc(sv)).append("\"}");
                n++;
                if (n >= 300) break;
            }
        }
        b.append("],\"count\":").append(n).append("}");
        writeOut(out, b.toString());
    }

    private void cmdData(String addrStr, String countStr, String out) throws Exception {
        Address addr;
        try { addr = currentProgram.getAddressFactory().getAddress(addrStr); }
        catch (Exception e) { writeOut(out, "{\"error\":\"bad address: " + esc(addrStr) + "\"}"); return; }
        int count = Integer.parseInt(countStr);
        byte[] buf = new byte[count];
        int got = currentProgram.getMemory().getBytes(addr, buf);
        StringBuilder hx = new StringBuilder();
        for (int i = 0; i < got; i++) {
            if (i > 0) hx.append(" ");
            hx.append(String.format("%02x", buf[i] & 0xff));
        }
        StringBuilder b = new StringBuilder();
        b.append("{\"addr\":\"").append(esc(addr.toString())).append("\",");
        b.append("\"count\":").append(got).append(",\"hex\":\"").append(esc(hx.toString())).append("\"}");
        writeOut(out, b.toString());
    }

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 2) { println("[GhidraQuery] usage: <command> <arg..> <outfile>"); return; }
        String cmd = args[0];
        if (cmd.equals("decompile"))      cmdDecompile(args[1], args[2]);
        else if (cmd.equals("func"))      cmdFunc(args[1], args[2]);
        else if (cmd.equals("struct"))    cmdStruct(args[1], args[2]);
        else if (cmd.equals("xref"))      cmdXref(args[1], args[2]);
        else if (cmd.equals("search"))    cmdSearch(args[1], args[2]);
        else if (cmd.equals("strings"))   cmdStrings(args[1], args[2]);
        else if (cmd.equals("data"))      cmdData(args[1], args[2], args[3]);
        else println("[GhidraQuery] unknown command: " + cmd);
    }
}
